File: repos/go-litellm/internal/core/streaming.go
package core

import (
	"bufio"
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/noizu-labs/go-litellm/internal/anthropic"
	"github.com/noizu-labs/go-litellm/internal/errx"
	"github.com/noizu-labs/go-litellm/internal/jsonx"
	"github.com/noizu-labs/go-litellm/internal/providers"
)

// StreamOpenAI streams a chat completion as OpenAI SSE (`data: …` + `[DONE]`).
func StreamOpenAI(w http.ResponseWriter, client HTTPDoer, p *Prepared) *errx.Error {
	return runStream(w, client, p, streamOpenAI{chunkID: "chatcmpl-" + providers.RandID(16), created: time.Now().Unix()})
}

// StreamAnthropic streams OpenAI-family upstream as Anthropic SSE events.
func StreamAnthropic(w http.ResponseWriter, client HTTPDoer, p *Prepared, requestedModel string) *errx.Error {
	return runStream(w, client, p, &streamAnthropic{model: requestedModel})
}

type emitter interface {
	preamble(http.ResponseWriter) error
	emit(http.ResponseWriter, providers.StreamChunk) error
	closing(http.ResponseWriter) error
}

type streamOpenAI struct {
	chunkID string
	created int64
	model   string
}

func (s streamOpenAI) preamble(http.ResponseWriter) error { return nil }

func (s streamOpenAI) emit(w http.ResponseWriter, chunk providers.StreamChunk) error {
	delta := map[string]any{}
	if chunk.Text != "" {
		delta["content"] = chunk.Text
	}
	if chunk.ToolUse != nil {
		delta["tool_calls"] = chunk.ToolUse
	}
	frame := map[string]any{
		"id":      s.chunkID,
		"object":  "chat.completion.chunk",
		"created": s.created,
		"model":   s.model,
		"choices": []any{
			map[string]any{
				"index":         chunk.Index,
				"delta":         delta,
				"finish_reason": chunk.FinishReason,
			},
		},
	}
	if chunk.Usage != nil {
		frame["usage"] = chunk.Usage
	}
	return writeSSEData(w, frame)
}

func (s streamOpenAI) closing(w http.ResponseWriter) error {
	_, err := io.WriteString(w, "data: [DONE]\n\n")
	flush(w)
	return err
}

type streamAnthropic struct {
	model        string
	finishReason any
	usage        any
}

func (s *streamAnthropic) preamble(w http.ResponseWriter) error {
	for _, f := range anthropic.StreamPreamble(s.model) {
		if _, err := io.WriteString(w, f); err != nil {
			return err
		}
	}
	flush(w)
	return nil
}

func (s *streamAnthropic) emit(w http.ResponseWriter, chunk providers.StreamChunk) error {
	if chunk.FinishReason != nil {
		s.finishReason = chunk.FinishReason
	}
	if chunk.Usage != nil {
		s.usage = chunk.Usage
	}
	text := chunk.Text
	if text == "" {
		text = chunk.Reasoning
	}
	if text == "" {
		return nil
	}
	_, err := io.WriteString(w, anthropic.StreamTextDelta(text))
	flush(w)
	return err
}

func (s *streamAnthropic) closing(w http.ResponseWriter) error {
	finish := ""
	if s.finishReason != nil {
		finish = stringifyAny(s.finishReason)
	}
	usage := jsonx.AsMap(s.usage)
	for _, f := range anthropic.StreamClosing(finish, usage) {
		if _, err := io.WriteString(w, f); err != nil {
			return err
		}
	}
	flush(w)
	return nil
}

func runStream(w http.ResponseWriter, client HTTPDoer, p *Prepared, em emitter) *errx.Error {
	if client == nil {
		client = DefaultClient
	}
	headers, e := p.Adapter.ValidateEnvironment(p.Req, map[string]string{})
	if e != nil {
		return e
	}
	if so, ok := em.(streamOpenAI); ok {
		so.model = p.Req.Model
		em = so
	}
	url := p.Adapter.CompleteURL(p.Req)
	payload := p.Adapter.TransformRequest(p.Req)
	raw, _ := json.Marshal(payload)
	httpReq, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(raw))
	if err != nil {
		return errx.New(502, "upstream stream failed: "+err.Error())
	}
	for k, v := range headers {
		httpReq.Header.Set(k, v)
	}
	httpReq.Header.Del("Accept-Encoding")
	resp, err := client.Do(httpReq)
	if err != nil {
		return errx.New(502, "upstream stream failed: "+err.Error())
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		b, _ := io.ReadAll(resp.Body)
		return p.Adapter.ErrorClass(resp.StatusCode, decodeBody(b), nil)
	}

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	w.WriteHeader(http.StatusOK)
	if err := em.preamble(w); err != nil {
		return nil
	}

	finished := false
	err = scanSSE(resp.Body, func(payload any, done bool) {
		if finished {
			return
		}
		if done {
			finished = true
			return
		}
		chunk := p.Adapter.ChunkParser(payload)
		if chunk.Done {
			finished = true
			return
		}
		_ = em.emit(w, chunk)
		if chunk.IsFinished {
			finished = true
		}
	})
	if err != nil {
		_ = writeSSEData(w, errx.New(502, "upstream stream failed: "+err.Error()).Body())
	}
	_ = em.closing(w)
	return nil
}

func scanSSE(r io.Reader, fn func(payload any, done bool)) error {
	br := bufio.NewReader(r)
	var event bytes.Buffer
	flushEvent := func() {
		data := extractData(event.String())
		event.Reset()
		if data == "" {
			return
		}
		if strings.TrimSpace(data) == "[DONE]" {
			fn(nil, true)
			return
		}
		var payload any
		if err := json.Unmarshal([]byte(data), &payload); err != nil {
			return
		}
		fn(payload, false)
	}
	for {
		line, err := br.ReadBytes('\n')
		if len(line) > 0 {
			if bytes.Equal(bytes.TrimSpace(line), []byte{}) {
				flushEvent()
			} else {
				event.Write(line)
			}
		}
		if err != nil {
			if event.Len() > 0 {
				flushEvent()
			}
			if err == io.EOF {
				return nil
			}
			return err
		}
	}
}

func extractData(event string) string {
	var parts []string
	for _, line := range strings.Split(event, "\n") {
		line = strings.TrimRight(line, "\r")
		if strings.HasPrefix(line, "data:") {
			parts = append(parts, strings.TrimSpace(strings.TrimPrefix(line, "data:")))
		}
	}
	return strings.Join(parts, "\n")
}

func writeSSEData(w http.ResponseWriter, v any) error {
	b, _ := json.Marshal(v)
	_, err := io.WriteString(w, "data: "+string(b)+"\n\n")
	flush(w)
	return err
}

func flush(w http.ResponseWriter) {
	if f, ok := w.(http.Flusher); ok {
		f.Flush()
	}
}

func stringifyAny(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	b, _ := json.Marshal(v)
	return strings.Trim(string(b), `"`)
}

File: repos/ex-litellm/lib/ex_litellm/core/streaming.ex
defmodule ExLiteLLM.Core.Streaming do
  @moduledoc """
  Server-Sent-Events streaming — litellm's `CustomStreamWrapper` +
  `BaseModelResponseIterator`.

  Upstream is driven with `Finch.stream/5` so we can inspect the HTTP status
  **before** committing a 200 SSE preamble. Groq (and other strict providers)
  often 400/404 with a JSON body; treating that as an Anthropic empty turn is
  how `run-claude with groq-pro` showed "models do not reply".
  """

  import Plug.Conn
  require Logger

  alias ExLiteLLM.Anthropic.Stream, as: AnthropicStream
  alias ExLiteLLM.Anthropic.Translate
  alias ExLiteLLM.Error
  alias ExLiteLLM.Providers.Adapter.Request

  @doc """
  Stream a chat completion to the client as OpenAI SSE.
  """
  @spec stream(Plug.Conn.t(), module(), Request.t(), map()) :: Plug.Conn.t()
  def stream(conn, adapter, %Request{} = req, upstream_body) do
    ExLiteLLM.Proxy.MetricsPlug.defer()
    ExLiteLLM.Proxy.MetricsPlug.tag(target: "native:#{req.provider}")

    with {:ok, headers} <- adapter.validate_environment(req, %{}) do
      state = %{
        conn: conn,
        adapter: adapter,
        req: req,
        chunk_id: "chatcmpl-" <> rand(),
        created: System.system_time(:second),
        buffer: "",
        finished: false,
        emit: :openai,
        sse_started: false,
        status: nil,
        error_buf: ""
      }

      case upstream_stream(adapter.get_complete_url(req), headers, upstream_body, req, state) do
        {:ok, final} ->
          final |> finalize_openai() |> ExLiteLLM.Proxy.MetricsPlug.finalize()

        {:error, %Error{} = e, %{sse_started: true} = st} ->
          send_error_frame(st.conn, e)

        {:error, %Error{} = e, _} ->
          send_error(conn, e)
      end
    else
      {:error, %Error{} = e} -> send_error(conn, e)
    end
  end

  @doc """
  Stream a chat completion to the client as **Anthropic SSE**.

  Content blocks are opened lazily (`thinking` / `text` / `tool_use`) so
  Groq gpt-oss reasoning and tool calls are visible to Claude Code.
  """
  @spec stream_anthropic(Plug.Conn.t(), module(), Request.t(), map(), String.t()) :: Plug.Conn.t()
  def stream_anthropic(conn, adapter, %Request{} = req, upstream_body, requested_model) do
    ExLiteLLM.Proxy.MetricsPlug.defer()
    ExLiteLLM.Proxy.MetricsPlug.tag(target: "native:#{req.provider}")

    with {:ok, headers} <- adapter.validate_environment(req, %{}) do
      state = %{
        conn: conn,
        adapter: adapter,
        req: req,
        buffer: "",
        finished: false,
        emit: :anthropic,
        anth: AnthropicStream.new(requested_model),
        sse_started: false,
        status: nil,
        error_buf: ""
      }

      case upstream_stream(adapter.get_complete_url(req), headers, upstream_body, req, state) do
        {:ok, final} ->
          final = flush_anthropic_close(final)
          ExLiteLLM.Proxy.MetricsPlug.finalize(final.conn)

        {:error, %Error{} = e, %{sse_started: true} = st} ->
          _ = chunk_frame(st.conn, Translate.stream_error(e.type, e.message))
          ExLiteLLM.Proxy.MetricsPlug.finalize(st.conn)

        {:error, %Error{} = e, _} ->
          send_error(conn, e)
      end
    else
      {:error, %Error{} = e} -> send_error(conn, e)
    end
  end

  # --- upstream via Finch (status before client SSE) ---

  defp upstream_stream(url, headers, body, %Request{litellm_params: lp}, state) do
    timeout = stream_timeout(lp)
    header_list = Enum.map(headers, fn {k, v} -> {to_string(k), to_string(v)} end)
    payload = if is_binary(body), do: body, else: Jason.encode!(body)
    request = Finch.build(:post, url, header_list, payload)

    case Finch.stream(request, ExLiteLLM.HTTP.finch(), state, &finch_step/2, receive_timeout: timeout) do
      {:ok, %{status: status} = acc} when status in 200..299 ->
        {:ok, acc}

      {:ok, %{status: status} = acc} when is_integer(status) ->
        {:error, adapter_error(acc.adapter, status, acc.error_buf), acc}

      {:ok, acc} ->
        {:error, Error.new(502, "upstream stream failed: no status", type: "api_error"), acc}

      {:error, exc} ->
        {:error, stream_exc(exc), state}
    end
  end

  defp finch_step({:status, status}, acc), do: %{acc | status: status}
  defp finch_step({:headers, _}, acc), do: acc
  defp finch_step({:trailers, _}, acc), do: acc

  defp finch_step({:data, data}, acc) do
    if acc.status in 200..299 do
      acc
      |> maybe_start_sse()
      |> consume(data)
    else
      %{acc | error_buf: acc.error_buf <> data}
    end
  end

  defp maybe_start_sse(%{sse_started: true} = acc), do: acc

  defp maybe_start_sse(%{emit: :anthropic} = acc) do
    conn =
      acc.conn
      |> put_resp_content_type("text/event-stream")
      |> put_resp_header("cache-control", "no-cache")
      |> send_chunked(200)

    {anth, frames} = AnthropicStream.preamble(acc.anth)
    conn = write_frames(conn, frames)
    %{acc | conn: conn, anth: anth, sse_started: true}
  end

  defp maybe_start_sse(acc) do
    conn =
      acc.conn
      |> put_resp_content_type("text/event-stream")
      |> put_resp_header("cache-control", "no-cache")
      |> send_chunked(200)

    %{acc | conn: conn, sse_started: true}
  end

  # Feed a raw byte blob into the SSE line parser.
  defp consume(state, data) do
    {events, rest} = split_sse(state.buffer <> data)
    state = %{state | buffer: rest}

    Enum.reduce(events, state, fn event, acc ->
      case parse_event(event) do
        :skip -> acc
        :done -> %{acc | finished: true}
        {:data, payload} -> handle_payload(acc, payload)
      end
    end)
  end

  defp handle_payload(state, payload) do
    if state.finished do
      state
    else
      chunk = state.adapter.chunk_parser(payload)
      emit(state, chunk)
    end
  end

  defp emit(state, :done), do: %{state | finished: true}

  defp emit(%{emit: :anthropic} = state, chunk) when is_map(chunk) do
    {anth, frames} = AnthropicStream.push(state.anth, chunk)
    conn = write_frames(state.conn, frames)

    %{
      state
      | anth: anth,
        conn: conn,
        finished: chunk[:is_finished] || state.finished
    }
  end

  defp emit(state, chunk) when is_map(chunk) do
    frame = "data: " <> Jason.encode!(openai_chunk_frame(state, chunk)) <> "\n\n"
    ExLiteLLM.Proxy.MetricsPlug.add_resp_bytes(byte_size(frame))

    case chunk(state.conn, frame) do
      {:ok, conn} -> %{state | conn: conn, finished: chunk.is_finished || state.finished}
      {:error, _} -> %{state | finished: true}
    end
  end

  defp flush_anthropic_close(state) do
    {anth, frames} = AnthropicStream.close(state.anth)
    %{state | anth: anth, conn: write_frames(state.conn, frames)}
  end

  defp write_frames(conn, frames) do
    Enum.reduce(frames, conn, fn frame, acc ->
      ExLiteLLM.Proxy.MetricsPlug.add_resp_bytes(byte_size(frame))

      case chunk(acc, frame) do
        {:ok, c} -> c
        {:error, _} -> acc
      end
    end)
  end

  defp chunk_frame(conn, frame) do
    ExLiteLLM.Proxy.MetricsPlug.add_resp_bytes(byte_size(frame))
    chunk(conn, frame)
  end

  # --- OpenAI chunk shaping ---

  defp openai_chunk_frame(state, chunk) do
    delta =
      %{}
      |> put_if(:content, blank_to_nil(chunk[:text] || chunk["text"]))
      |> put_if(:tool_calls, chunk[:tool_use] || chunk["tool_use"])
      |> put_if(:reasoning, blank_to_nil(chunk[:reasoning] || chunk["reasoning"]))

    %{
      id: state.chunk_id,
      object: "chat.completion.chunk",
      created: state.created,
      model: state.req.model,
      choices: [
        %{
          index: chunk[:index] || 0,
          delta: delta,
          finish_reason: chunk[:finish_reason]
        }
      ]
    }
    |> maybe_usage(chunk[:usage])
  end

  defp maybe_usage(frame, nil), do: frame
  defp maybe_usage(frame, usage), do: Map.put(frame, :usage, usage)

  defp finalize_openai(%{error: %Error{} = e, conn: conn}), do: send_error_frame(conn, e)

  defp finalize_openai(state) do
    {:ok, conn} = chunk(state.conn, "data: [DONE]\n\n")
    conn
  end

  # --- SSE parsing ---

  defp split_sse(buffer) do
    parts = String.split(buffer, ~r/\r?\n\r?\n/)

    case Enum.reverse(parts) do
      [last | rest] -> {Enum.reverse(rest), last}
      [] -> {[], ""}
    end
  end

  defp parse_event(event) do
    data =
      event
      |> String.split(~r/\r?\n/)
      |> Enum.filter(&String.starts_with?(&1, "data:"))
      |> Enum.map(&String.trim_leading(String.replace_prefix(&1, "data:", "")))
      |> Enum.join("\n")

    cond do
      data == "" -> :skip
      String.trim(data) == "[DONE]" -> :done
      true -> decode_data(data)
    end
  end

  defp decode_data(data) do
    case Jason.decode(data) do
      {:ok, payload} -> {:data, payload}
      {:error, _} -> :skip
    end
  end

  defp adapter_error(adapter, status, buf) do
    body =
      case Jason.decode(buf) do
        {:ok, decoded} -> decoded
        _ -> buf
      end

    adapter.get_error_class(status, body, %{})
  end

  defp send_error(conn, %Error{} = e) do
    conn
    |> put_resp_content_type("application/json")
    |> send_resp(e.status, Jason.encode!(anthropic_or_openai(conn, e)))
  end

  defp anthropic_or_openai(conn, %Error{} = e) do
    if String.contains?(conn.request_path || "", "/messages") do
      %{"type" => "error", "error" => %{"type" => e.type, "message" => e.message}}
    else
      Error.to_body(e)
    end
  end

  defp send_error_frame(conn, %Error{} = e) do
    {:ok, conn} = chunk(conn, "data: " <> Jason.encode!(Error.to_body(e)) <> "\n\n")
    {:ok, conn} = chunk(conn, "data: [DONE]\n\n")
    conn
  end

  defp put_if(map, _key, nil), do: map
  defp put_if(map, key, value), do: Map.put(map, key, value)

  defp blank_to_nil(""), do: nil
  defp blank_to_nil(v), do: v

  defp stream_timeout(%{"stream_timeout" => t}) when is_number(t), do: round(t * 1000)
  defp stream_timeout(%{"timeout" => t}) when is_number(t), do: round(t * 1000)
  defp stream_timeout(_), do: 600_000

  defp stream_exc(%{__struct__: _} = exc),
    do: Error.new(502, "upstream stream failed: #{Exception.message(exc)}", type: "api_error")

  defp stream_exc(other),
    do: Error.new(502, "upstream stream failed: #{inspect(other)}", type: "api_error")

  defp rand, do: 16 |> :crypto.strong_rand_bytes() |> Base.url_encode64(padding: false)
end
